### PR TITLE
Add int128 and supported functions

### PR DESCRIPTION
**Pull request by Ashleyhx**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,10 @@ add_subdirectory(processor)
 add_subdirectory(storage)
 add_subdirectory(transaction)
 
-add_library(kuzu STATIC ${ALL_OBJECT_FILES})
+add_library(kuzu STATIC ${ALL_OBJECT_FILES}
+        include/common/types/int128_t.h
```
Comment (acquamarin) [inline src/CMakeLists.txt:22]: remove

File: src/common/types/int128_t.cpp [hunk 2]
```diff
@@ -0,0 +1,768 @@
+#include "common/types/int128_t.h"
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:1]: My general comment on this file is:
1. Use camel case for function names.
2. Remove any functions that are unused

File: src/include/c_api/kuzu.h [hunk 3]
```diff
@@ -899,6 +900,8 @@ KUZU_C_API uint32_t kuzu_value_get_uint32(kuzu_value* value);
  * @param value The value to return.
  */
 KUZU_C_API uint64_t kuzu_value_get_uint64(kuzu_value* value);
+
+//KUZU_C_API int128_t kuzu_value_get_int128(kuzu_value* value);
```
Comment (acquamarin) [inline src/include/c_api/kuzu.h:904]: Guess, you need this api

File: src/include/function/cast/numeric_cast.h [hunk 4]
```diff
@@ -232,6 +232,83 @@ inline uint8_t castStringToNum(const char* input, uint64_t len, const common::Lo
     simpleIntegerCast<uint8_t, false>(input, len, result, type);
     return result;
 }
+//==============================================================================
```
Comment (acquamarin) [inline src/include/function/cast/numeric_cast.h:235]: remove the comment since it is trivial

File: src/include/function/cast/numeric_cast.h [hunk 5]
```diff
@@ -232,6 +232,83 @@ inline uint8_t castStringToNum(const char* input, uint64_t len, const common::Lo
...
+    }
+
+    template<class T, bool NEGATIVE>
+    static bool finalize(T& result) {
+        return result.flush();
+    }
+};
+
+static bool trySimpleInt128Cast(const char* input, uint64_t len, common::int128_t& result) {
+    Int128CastData data{};
```
Comment (acquamarin) [inline src/include/function/cast/numeric_cast.h:294]: Move the implementation to .cpp

File: src/include/function/cast/numeric_limits.h [hunk 6]
```diff
@@ -64,5 +76,6 @@ constexpr uint64_t NumericLimits<double>::digits() {
     return 250;
 }
 
+
```
Comment (acquamarin) [inline src/include/function/cast/numeric_limits.h:79]: remove newline

File: test/test_files/tinysnb/projection/single_label.test [hunk 7]
```diff
@@ -5,673 +5,12 @@
 
 -CASE ProjectionSingleLabel
 
--LOG ReturnLiteral3
```
Comment (acquamarin) [inline test/test_files/tinysnb/projection/single_label.test:8]: Why delete those tests?

**Review by acquamarin — COMMENTED** [review]
Can you add some test cases to cast_error.test where the casting fails?

File: src/function/vector_list_functions.cpp [hunk 8]
```diff
@@ -468,6 +475,9 @@ std::unique_ptr<FunctionBindData> ListReverseSortVectorFunction::bindFunc(
     case LogicalTypeID::UINT8: {
         getExecFunction<uint8_t>(arguments, vectorFunctionDefinition->execFunc);
     } break;
+    case LogicalTypeID::INT128: {
+        getExecFunction<int128_t>(arguments, vectorFunctionDefinition->execFunc);
```
Comment (acquamarin) [inline src/function/vector_list_functions.cpp:479]: Add tests for int128 related list functions

File: src/include/common/int128_t.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <cstdint>
+
+#include "common/api.h"
+#include "common/string_utils.h"
+
+namespace kuzu::common {
+
+struct KUZU_API int128_t {
```
Comment (acquamarin) [inline src/include/common/int128_t.h:10]: Our convension it to write KUZU_API before stuct/class declaration

File: src/include/common/int128_t.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    uint64_t low;
+    int64_t high;
+
+    int128_t() = default;
+    int128_t(int64_t value); // NOLINT: Allow implicit conversion from `int64_t`
```
Comment (acquamarin) [inline src/include/common/int128_t.h:15]: BTW: Where are we going to use this implicit cast?

File: src/include/function/arithmetic/arithmetic_functions.h [hunk 11]
```diff
@@ -89,20 +100,35 @@ struct Abs {
     }
 };
 
+template<>
+inline void Abs::operation(common::int128_t& input, common::int128_t& result) {
```
Comment (acquamarin) [inline src/include/function/arithmetic/arithmetic_functions.h:104]: Since you have already implemented operations for int128_t, do you still need to specialize hugeint?

File: src/parser/transform/transform_expression.cpp [hunk 12]
```diff
@@ -579,8 +579,15 @@ std::string Transformer::transformPropertyKeyName(CypherParser::OC_PropertyKeyNa
 std::unique_ptr<ParsedExpression> Transformer::transformIntegerLiteral(
     CypherParser::OC_IntegerLiteralContext& ctx) {
     auto text = ctx.DecimalInteger()->getText();
-    auto value =
-        std::make_unique<Value>(function::castStringToNum<int64_t>(text.c_str(), text.length()));
+    function::IntegerCastData<int64_t> data{};
+    auto len = text.length();
+    if (tryIntegerCast<function::IntegerCastData<int64_t>, true>(text.c_str(), len, data)) {
```
Comment (acquamarin) [inline src/parser/transform/transform_expression.cpp:584]: This line doesn't compile on mac, guess you need to cast len to write `(uint64_t&)len`

File: src/common/types/int128_t.cpp [hunk 13]
```diff
@@ -0,0 +1,766 @@
...
+        uint64_t low = input.low;
+        while (low) {
+            low >>= 1;
+            result++;
+        }
+    }
+    return result;
+}
+
+static bool PositiveInt128IsBitSet(int128_t input, uint8_t bit) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:77]: Use camel case for all functions/vars

File: src/common/types/int128_t.cpp [hunk 14]
```diff
@@ -0,0 +1,766 @@
...
+    }
+    return div_result;
+}
+
+int128_t Int128_t::Div(int128_t lhs, int128_t rhs) {
+    int128_t remainder{};
+    return DivMod(lhs, rhs, remainder);
+}
+
+int128_t Int128_t::Mod(int128_t lhs, int128_t rhs) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:350]: Remove those functions if not used

File: src/common/types/int128_t.cpp [hunk 15]
```diff
@@ -0,0 +1,766 @@
...
+    return true;
+}
+
+template<>
+bool Int128_t::TryCastTo(float value, int128_t& result) {
+    return TryCastTo(double(value), result);
+}
+
+template<class REAL_T>
+bool CastFloatingToInt128(REAL_T value, int128_t& result) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:527]: Ditto remove

**Review by acquamarin — COMMENTED** [review]
Can you also fix the CI and compilation issue on mac?

File: src/include/function/arithmetic/arithmetic_functions.h (hunk 11 above)
Comment (Ashleyhx) [reply src/include/function/arithmetic/arithmetic_functions.h:104]: yes here, since for abs we use c++'s` std::abs` which does not support int128, so we need to specialize hugeint, but for other functions like add, sub we don't need to specialize.

File: src/include/common/int128_t.h (hunk 10 above)
Comment (Ashleyhx) [reply src/include/common/int128_t.h:15]: `createDefaultValue` in value.cpp 
```
case LogicalTypeID::INT128:
        return Value(int128_t(0));
```

File: CMakeLists.txt [hunk 16]
```diff
@@ -224,7 +224,7 @@ else()
         find_library(ARROW_PATH arrow HINTS ${ARROW_INSTALL}/lib ${ARROW_INSTALL}/lib64)
     endif()
 
-    add_library(arrow_deps STATIC IMPORTED)
+    add_library(arrow_deps STATIC IMPORTED tools/python_api/src_cpp/py_int128.h)
```
Comment (acquamarin) [inline CMakeLists.txt:227]: Why do we need this?

File: src/c_api/value.cpp [hunk 17]
```diff
@@ -95,6 +95,13 @@ kuzu_value* kuzu_value_create_uint64(uint64_t val_) {
     return c_value;
 }
 
+kuzu_value* kuzu_value_create_int128(kuzu_int128_t val_) {
```
Comment (acquamarin) [inline src/c_api/value.cpp:98]: I will let @mewim to review api related stuff.

File: src/common/types/int128_t.cpp [hunk 18]
```diff
@@ -0,0 +1,650 @@
...
+#include <limits>
+
+#include "common/int128_t.h"
+#include "common/types/ku_string.h"
+
+namespace kuzu {
+namespace common {
+
+//===============================================================================================
+// String Convert Helper
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:13]: Remove this comment

File: src/common/types/int128_t.cpp [hunk 19]
```diff
@@ -0,0 +1,650 @@
...
+    }
+    if (result.empty()) {
+        result = "0";
+    }
+    return negative ? "-" + result : result;
+}
+
+//===============================================================================================
+// Arithmetic operators: ADD SUB
+//================================================================================================
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:136]: The function name is self-explanatory, you can remove the comment

File: src/common/types/int128_t.cpp [hunk 20]
```diff
@@ -0,0 +1,650 @@
...
+        }
+        lhs.high = lhs.high + rhs.high + overflow;
+    } else {
+        if (lhs.high < INT64_MIN - rhs.high - overflow) {
+            return false;
+        }
+        lhs.high = lhs.high + rhs.high + overflow;
+    }
+    lhs.low += rhs.low;
+    if (lhs.high == INT64_MIN && lhs.low == 0) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:151]: My general comment is that, we should add tests to cover the edge case.

File: src/common/types/int128_t.cpp [hunk 21]
```diff
@@ -0,0 +1,650 @@
...
+
+bool Int128_t::subInPlace(int128_t& lhs, int128_t rhs) {
+    int underflow = lhs.low - rhs.low > lhs.low;
+    if (rhs.high >= 0) {
+        if (lhs.high < INT64_MIN + rhs.high + underflow) {
+            return false;
+        }
+        lhs.high = lhs.high - rhs.high - underflow;
+    } else {
+        if (lhs.high > INT64_MIN && lhs.high - 1 >= INT64_MAX + rhs.high + underflow) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:165]: Ditto add edge case

File: src/common/types/int128_t.cpp [hunk 22]
```diff
@@ -0,0 +1,650 @@
...
+    lhs.low -= rhs.low;
+    if (lhs.high == INT64_MIN && lhs.low == 0) {
+        return false;
+    }
+    return true;
+}
+
+int128_t Int128_t::Add(int128_t lhs, const int128_t rhs) {
+    if (!addInPlace(lhs, rhs)) {
+        throw std::overflow_error("INT128 is out of range: cannot add.");
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:179]: Use // LCOV_EXCL_START and // LCOV_EXCL_END to let codecov ignore the coverage check for error message

File: src/common/types/int128_t.cpp [hunk 23]
```diff
@@ -0,0 +1,650 @@
...
+    }
+    return lhs;
+}
+
+int128_t Int128_t::Sub(int128_t lhs, const int128_t rhs) {
+    if (!subInPlace(lhs, rhs)) {
+        throw std::overflow_error("INT128 is out of range: cannot subtract.");
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:186]: Instead of throwing c++ execptions, we should throw kuzu internal exception. Add a new exception type: overflowException

File: src/common/types/int128_t.cpp [hunk 24]
```diff
@@ -0,0 +1,650 @@
...
+    if (lhs_negative ^ rhs_negative) {
+        negateInPlace(div_result);
+    }
+    if (lhs_negative) {
+        negateInPlace(remainder);
+    }
+    return div_result;
+}
+
+int128_t Int128_t::Div(int128_t lhs, int128_t rhs) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:343]: I think hugeint/hugeint will probably use this function.

File: src/common/types/int128_t.cpp [hunk 25]
```diff
@@ -0,0 +1,650 @@
...
+}
+
+// inplace arithmetic operators
+int128_t& int128_t::operator+=(const int128_t& rhs) {
+    if (!Int128_t::addInPlace(*this, rhs)) {
+        throw std::overflow_error("INT128 is out of range: cannot add in place.");
+    }
+    return *this;
+}
+int128_t& int128_t::operator-=(const int128_t& rhs) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:617]: Do you need these inplace operators?

File: src/common/types/int128_t.cpp [hunk 26]
```diff
@@ -0,0 +1,650 @@
...
+int128_t::operator bool() const {
+    return *this != 0;
+}
+
+template<class T>
+static T NarrowCast(const int128_t& input) {
+    return static_cast<T>(input.low);
+}
+
+int128_t::operator int64_t() const {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:645]: I think this is for implicit casting, you also don't need it

File: test/test_files/tinysnb/function/comparison.test [hunk 27]
```diff
@@ -316,4 +339,4 @@ False
 -LOG StructNotEqualWithNull
 -STATEMENT RETURN {a: 2, b: null} <> {a: 2, b: null}
 ---- 1
-False
+False
```
Comment (acquamarin) [inline test/test_files/tinysnb/function/comparison.test:342]: Always add a newline at end of file.

**Review by acquamarin — COMMENTED** [review]
Good job! One general comment is that we should increase the test coverage for edge cases.
I will let @mewim review the API.

**Review by mewim — COMMENTED** [review]
I went through the API related changes. The changes for Python and Node.js APIs LGTM.

For C API, the binding is not very user-friendly. The users will not understand `val_.low, val_.high` until they look into our internal implementation. I suggest to at least bind two more functions: `kuzu_int128_t kuzu_int128_t_from_string(const char* str)` and `char* kuzu_int128_t_to_string(kuzu_int128_t val)`, so that the user can convert `kuzu_int128_t` from/to strings without understanding the internal implementation.

For Java and Rust APIs, I see that only the type enums have been added, but no actual binding implementations. It is fine to implement these bindings as separate PRs, but an issue should be created for them so we don't miss them for the next release.

**Comment by mewim** [comment]
Also, we should change the title of this PR so it looks better in our release change log.

File: src/common/types/int128_t.cpp [hunk 28]
```diff
@@ -0,0 +1,633 @@
...
+    return TryCastInt128Template<uint32_t>(input, result);
+}
+
+template<>
+bool Int128_t::tryCast(int128_t input, uint64_t& result) {
+    return TryCastInt128Template<uint64_t>(input, result);
+}
+
+template<>
+bool Int128_t::tryCast(int128_t input, int128_t& result) {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:413]: When do we need to cast an int128_t value to an int128_t, seems unnecessary

File: src/common/types/int128_t.cpp [hunk 29]
```diff
@@ -0,0 +1,633 @@
...
+
+bool int128_t::operator!=(const int128_t& rhs) const {
+    return Int128_t::notEquals(*this, rhs);
+}
+
+bool int128_t::operator>(const int128_t& rhs) const {
+    return Int128_t::greaterThan(*this, rhs);
+}
+
+bool int128_t::operator>=(const int128_t& rhs) const {
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:566]: Add test case for int128_t comparison

File: src/common/types/int128_t.cpp [hunk 30]
```diff
@@ -0,0 +1,633 @@
...
+}
+
+int128_t int128_t::operator-() const {
+    return Int128_t::negate(*this);
+}
+
+// inplace arithmetic operators
+int128_t& int128_t::operator+=(const int128_t& rhs) {
+    if (!Int128_t::addInPlace(*this, rhs)) {
+        throw common::OverflowException("INT128 is out of range: cannot add in place.");
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:605]: Use codecov instructions to ignore the coverage check for this exception.
BTW: When do we need to call inplace add?

File: src/common/types/int128_t.cpp [hunk 31]
```diff
@@ -0,0 +1,633 @@
...
+    *this = Int128_t::Mul(*this, rhs);
+    return *this;
+}
+
+bool int128_t::operator!() const {
+    return *this == 0;
+}
+
+int128_t::operator bool() const {
+    return *this != 0;
```
Comment (acquamarin) [inline src/common/types/int128_t.cpp:620]: I don't think you need these two operators

File: src/common/types/types.cpp [hunk 32]
```diff
@@ -36,6 +37,8 @@ std::string PhysicalTypeUtils::physicalTypeToString(PhysicalTypeID physicalType)
         return "UINT16";
     case PhysicalTypeID::UINT8:
         return "UINT8";
+    case PhysicalTypeID::INT128:
+        return "INT128";
```
Comment (acquamarin) [inline src/common/types/types.cpp:41]: Guess, we should ignore the coverage check for the `PhysicalTypeUtils::physicalTypeToString` function.

File: src/common/types/types.cpp [hunk 33]
```diff
@@ -709,17 +723,18 @@ std::vector<LogicalType> LogicalTypeUtils::getAllValidComparableLogicalTypes() {
...
         LogicalType{LogicalTypeID::UINT16}, LogicalType{LogicalTypeID::UINT8},
-        LogicalType{LogicalTypeID::DOUBLE}, LogicalType{LogicalTypeID::FLOAT},
-        LogicalType{LogicalTypeID::DATE}, LogicalType{LogicalTypeID::TIMESTAMP},
-        LogicalType{LogicalTypeID::INTERVAL}, LogicalType{LogicalTypeID::BLOB},
-        LogicalType{LogicalTypeID::STRING}, LogicalType{LogicalTypeID::SERIAL}};
+        LogicalType{LogicalTypeID::INT128}, LogicalType{LogicalTypeID::DOUBLE},
+        LogicalType{LogicalTypeID::FLOAT}, LogicalType{LogicalTypeID::DATE},
+        LogicalType{LogicalTypeID::TIMESTAMP}, LogicalType{LogicalTypeID::INTERVAL},
+        LogicalType{LogicalTypeID::BLOB}, LogicalType{LogicalTypeID::STRING},
+        LogicalType{LogicalTypeID::SERIAL}};
```
Comment (acquamarin) [inline src/common/types/types.cpp:730]: Add int128_t to validTypes as well.

File: src/function/vector_cast_functions.cpp [hunk 34]
```diff
@@ -109,6 +111,10 @@ void VectorCastFunction::bindImplicitCastFunc(
         bindImplicitNumericalCastFunc<uint64_t, CastToUInt64>(sourceTypeID, func);
         return;
     }
+    case LogicalTypeID::INT128: {
+        bindImplicitNumericalCastFunc<int128_t, CastToInt128>(sourceTypeID, func);
```
Comment (acquamarin) [inline src/function/vector_cast_functions.cpp:115]: When do we need to cast int128_t to 128_t?

File: src/include/common/exception/overflow.h [hunk 35]
```diff
@@ -0,0 +1,14 @@
...
+namespace kuzu {
+namespace common {
+
+class OverflowException : public Exception {
+public:
+    explicit OverflowException(const std::string& msg) : Exception("Overflow exception: " + msg) {}
+};
+
+} // namespace common
+} // namespace kuzu
```
Comment (acquamarin) [inline src/include/common/exception/overflow.h:14]: newline

File: src/include/common/int128_t.h [hunk 36]
```diff
@@ -0,0 +1,59 @@
...
+    int128_t operator/(const int128_t& rhs) const;
+    int128_t operator%(const int128_t& rhs) const;
+    int128_t operator-() const;
+
+    // inplace arithmetic operators
+    int128_t& operator+=(const int128_t& rhs);
+    int128_t& operator*=(const int128_t& rhs);
+
+    // boolean operators
+    explicit operator bool() const;
```
Comment (acquamarin) [inline src/include/common/int128_t.h:47]: Double check whether we need these operators or not.

File: src/include/function/aggregate/avg.h [hunk 37]
```diff
@@ -80,7 +81,12 @@ struct AvgFunction {
     static void finalize(uint8_t* state_) {
         auto state = reinterpret_cast<AvgState*>(state_);
         if (!state->isNull) {
-            state->avg = state->sum / (double_t)state->count;
+            if constexpr (std::is_same_v<T, common::int128_t>) {
```
Comment (acquamarin) [inline src/include/function/aggregate/avg.h:84]: Use template specialization instead of `if constexpr` to improve readability.
Also add a test case on avg function on huge int.

File: src/include/function/arithmetic/arithmetic_functions.h [hunk 38]
```diff
@@ -64,6 +64,17 @@ inline void Modulo::operation(int64_t& left, int64_t& right, int64_t& result) {
     result = left % right;
 }
 
+template<>
+inline void Modulo::operation(
+    common::int128_t& left, common::int128_t& right, common::int128_t& result) {
+    if (right == 0) {
+        // According to c++ standard, only INT128 % 0(INT128) is undefined. (eg. DOUBLE % 0(INT128)
+        // and INT128 % 0.0(DOUBLE) are well-defined).
+        throw common::RuntimeException("Modulo by zero.");
```
Comment (acquamarin) [inline src/include/function/arithmetic/arithmetic_functions.h:73]: Use lcov instruction to ignore coverage check for this.

File: src/include/function/cast/cast_utils.h [hunk 39]
```diff
@@ -74,7 +76,7 @@ static bool integerCastLoop(const char* input, uint64_t len, T& result) {
     return pos > start_pos; // false if no digits "" or "-"
 }
 
-template<typename T, bool IS_SIGNED = true>
+template<class T, bool IS_SIGNED = true, class OP = IntegerCastOperation>
```
Comment (acquamarin) [inline src/include/function/cast/cast_utils.h:79]: We prefere to use typename T instead of class T

File: src/include/function/cast/cast_utils.h [hunk 40]
```diff
@@ -92,15 +94,90 @@ static bool tryIntegerCast(const char* input, uint64_t& len, T& result) {
...
     }
 
     // not allow leading 0
     if (len > 1 && *input == '0') {
         return false;
     }
+    return integerCastLoop<T, false, false, OP>(input, len, result);
+}
+
+// INT128 Cast Helper
```
Comment (acquamarin) [inline src/include/function/cast/cast_utils.h:107]: This comment looks redudant.

**Review by acquamarin — APPROVED** [review]
My general comment is to improve test coverage as much as you can.

File: src/common/types/int128_t.cpp (hunk 30 above)
Comment (Ashleyhx) [reply src/common/types/int128_t.cpp:605]: in list_sum_function case for int128

File: src/include/common/int128_t.h (hunk 36 above)
Comment (Ashleyhx) [reply src/include/common/int128_t.h:47]: inplace mult is used in vector_list_function

File: src/include/c_api/kuzu.h [hunk 41]
```diff
@@ -882,6 +894,19 @@ KUZU_C_API uint32_t kuzu_value_get_uint32(kuzu_value* value);
...
 KUZU_C_API uint64_t kuzu_value_get_uint64(kuzu_value* value);
+/**
+ * @brief Returns the int128 value of the given value. The value must be of type INT128.
+ * @param value The value to return.
+ */
+KUZU_C_API kuzu_int128_t kuzu_value_get_int128(kuzu_value* value);
+/**
+ * @brief convert a string to int128 value.
+ */
+KUZU_C_API kuzu_int128_t kuzu_int128_t_from_string(char* str);
```
Comment (mewim) [inline src/include/c_api/kuzu.h:905]: The function signature here should be `const char* str`

**Review by mewim — APPROVED** [review]
